### ISSUE TITLE
[FSDP2][DTensor] numeric bug for DTensor + python float in gradient clipping

### ISSUE THREAD

**Issue by weifengpy** [pytorch/pytorch#149768]
### 🐛 Describe the bug

```
python3.12/site-packages/torch/nn/utils/clip_grad.py", line 155, in _clip_grads_with_norm_
[rank1]:     clip_coef = max_norm / (total_norm + 1e-6)
```

for DTensor + 1e-6, each rank adds 1e-6 to local tensor, instead of adding once

this was reported by others as well

### Versions

na

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci @zhaojuanmao @mrshenli @rohan-varma @chauhang @mori360 @kwen2501 @c-p-i-o

**Comment by weifengpy** [comment]
@leonardo0lyj are you experiencing the same issue?

**Comment by awgu** [comment]
cc: @tianyu-l  how should adding to a partial tensor work in sharding prop? My intuition is that we should divide the added scalar here by the mesh dim size before adding.

**Comment by RishabhSaini** [comment]
Can this be assigned to me? I plan on working on this, will have a PR soon

**Comment by RishabhSaini** [comment]
This is the logic that dtensors tend to follow when a Tensor -> Sharded Tensor -> Partial Norm -> Reduced Placement
Example for nnodes=1 and nproc_per_node=2 and Tensor [[1, 1], [2, 2]]

| Rank   | dtensor       | `_NormPartial(reduce_op=Sum, norm_type=2.0)` | `_reduce_values` -> Placement type = Replicate                                | `_post_reduce_transform`                         |
|--------|---------------|-----------------------------------------------|--------------------------------------------------|--------------------------------------------------|
| Rank 0 | [1, 1]        | local_norm = √2                         |   (√2)² + (2√2)² = 2 + 8 = 10               | √10                                                       |
| Rank 1 | [2, 2]        | local_norm = 2√2                       | (√2)² + (2√2)² = 2 + 8 = 10                 | √10                                                       |

This validates that mathematically the above should be equivalent to:
torch.linalg.vector_norm(torch.tensor([[1.0, 1.0], [2.0, 2.0]]))

However below tells us the minute difference between both:

```
>>> import torch
>>> print(f"{(torch.linalg.vector_norm(torch.tensor([1.0, 1.0]))**2.0 + torch.linalg.vector_norm(torch.tensor([2.0, 2.0]))**2.0)**(1/2):.10f}")
3.1622774601
>>> print(f"{torch.linalg.vector_norm(torch.tensor([[1.0, 1.0], [2.0, 2.0]])):.10f}")
3.1622776985
```

**Comment by RishabhSaini** [comment]
A more appropriate solution will change the `_NormPartial` to a `_SquaredNormPartial`. 
Essentially within `_get_total_norm` for a sharded dtensor will get the _SquaredNormPartial for each shard. When the _SquaredNormPartial is converted to a Replicate, the `_pre_reduce_transform()` will be skipped.

Example for nnodes=1 and nproc_per_node=2 and Tensor [[1, 1], [2, 2]]

| Rank   | dtensor       | `_SqauredNormPartial(reduce_op=Sum, norm_type=2.0)` | `_reduce_values` -> Placement type = Replicate                                | `_post_reduce_transform`                         |
|--------|---------------|-----------------------------------------------|--------------------------------------------------|--------------------------------------------------|
| Rank 0 | [1, 1]        | local_norm = torch.sum([1, 1]**2) = 2                         |   2 + 8 = 10               | √10                                                       |
| Rank 1 | [2, 2]        | local_norm = torch.sum([2, 2]**2) = 8                      | 2 + 8 = 10                 | √10                                                       |

This will fuse the operations of torch.sum(shardedTensor**norm_type), preventing any rounding errors.
This will require a much bigger change.

@fduwjj @bdhirsh @awgu @weifengpy What are your thoughts on this implementation?
Should I proceed with this?

**Comment by vadimkantorov** [comment]
Regarding squared norm, there was a discussion for a new `torch.powersum` frontend method:
- https://github.com/pytorch/pytorch/issues/28119#issuecomment-575255083

**Comment by tianyu-l** [comment]
@RishabhSaini It sounds to me that you are solving a harder challenging regarding floating point limitation.

The original post was about a bug on `_NormPartial` + scalar results in wrong placements, I believe.

**Comment by leonardo0lyj** [comment]
I agree with @tianyu-l @weifengpy @awgu.

Current `clip_grad_norm(dtensors)` gives wrong results, because the result is proportional to number of devices (detrimental in hyper-scale training):

|  	| DTensor(_NormPartial) 	| Add 1E-6 	| Redistribute(Replicate) 	|
|---	|---	|---	|---	|
| Rank0 	| 0.5 	| 0.5+1E-6 	| 1+2E-6 	|
| Rank1 	| 0.5 	| 0.5+1E-6 	| 1+2E-6 	|


Ideally, `clip_grad_norm(dtensors)` should give single-device-equal results, regardless number of devices:
|  	| DTensor(_NormPartial) 	| Redistribute(Replicate) 	| Add 1E-6 	|
|---	|---	|---	|---	|
| Rank0 	| 0.5 	| 1 	| 1+E-6 	|
| Rank1 	| 0.5 	| 1 	| 1+E-6 	|

Solution-A: do what @awgu suggest -- pre-divide 1E6 by number of devices

Solution-B: reorder the Redistribute and Add

cc @jiachengyang

**Comment by kv9699** [comment]
I think a better way to solve it may be to throw an error when adding `Partial` with a scalar unless users explicitly use something like `implicit_replication`. And in that case, just trade the scalar as a replicate dtensor and trigger `dtensor.redistribute` automatically.

**Comment by anshul-si** [comment]
Resolved in https://github.com/pytorch/pytorch/pull/170030
